### PR TITLE
ENH: Add ods writer

### LINKED ISSUE #27222: Wish: Write support for Open Document Spreadsheet (ODS)
Ni!

This is a follow up to #2311 which has been closed after read support was implemented in  #25427 .

People whose work flow involve outputting ODS files would benefit from write support.

.~´

### PR DESCRIPTION
**Pull request by roberthdevries**
- [x] closes #27222
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] add support for startrow and startcol parameters

### CODE REVIEW

File: pandas/io/excel/_odfreader.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from typing import List
 
+import numpy as np
```
Comment (WillAyd) [inline pandas/io/excel/_odfreader.py]: Is there a reason for the changes to the reader here? If solving a separate issue should do in a separate PR

File: pandas/tests/io/excel/test_writers.py [hunk 2]
```diff
@@ -1166,7 +1173,9 @@ def test_bytes_io(self, engine):
         writer.save()
 
         bio.seek(0)
-        reread_df = pd.read_excel(bio, index_col=0)
+        if engine != "odf":
+            engine = None
```
Comment (WillAyd) [inline pandas/tests/io/excel/test_writers.py]: Are you setting to None because this test doesn't work for odf? If so you should just `pytest.xfail()` if not implemented or `pytest.skip()` if not applicable

File: pandas/io/excel/_odswriter.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            wks = Table(name=sheet_name)
+            self.sheets[sheet_name] = wks
+
+        if _validate_freeze_panes(freeze_panes):
+            self._create_freeze_panes(sheet_name, freeze_panes)
+
+        rows: DefaultDict = defaultdict(TableRow)
+        col_count: DefaultDict = defaultdict(int)
+
+        for cell in sorted(cells, key=lambda cell: (cell.row, cell.col)):
```
Comment (WillAyd) [inline pandas/io/excel/_odswriter.py]: Is there a reason for the sorted call here? We don't do this for other writers?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Cool thanks for taking a stab at this. I realize a lot of this is copy / paste; we may need to think about defining a better base class (not sure yet if a pre-cursor or follow up)

File: pandas/io/excel/_odswriter.py (hunk 3 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: In order to add cells and rows to the odf file we have to add stuff in order. I am basically building up an XML file here. Adding cells/rows out of order would cause a sheet with data in the wrong cells

File: pandas/io/excel/_odfreader.py (hunk 1 above)
Comment (roberthdevries) [reply pandas/io/excel/_odfreader.py]: This came up when doing roundtrip tests. I had to fix this one or else every 0 in the sheet would be removed.

File: pandas/tests/io/excel/test_writers.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_writers.py]: No, when you set engine to None, it uses the default engine (xlrd IIRC)

Comment (WillAyd) [reply pandas/tests/io/excel/test_writers.py]: Yea so my point is we shouldn't be changing the fixture like this. Should either be an xfail or skip depending on comment above

File: pandas/io/excel/_odswriter.py (hunk 3 above)
Comment (WillAyd) [reply pandas/io/excel/_odswriter.py]: Right but aren't these already in the order you need? Noted on comment but still not clear on why this needs to be done differently for odf

Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: Most of the time yes, but there were a couple of tests where this was not the case (I forgot which one, but I think this had something to do with adding an index)

File: pandas/tests/io/excel/test_writers.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_writers.py]: No tests are skipped, only the right engine is chosen. When having a byte iostream you have to explicitly pass the engine, or if you pass `None` it uses the default engine (which is `xlrd` IIRC).

**Comment by roberthdevries** [comment]
@WillAyd I am not sure where "the lot of copy paste" comes from. It is true that I started off from `_xlwt.py`, but there are just 20 unmodified lines of the 321. Most of which are the standard methods required for each engine.

File: pandas/tests/io/excel/test_writers.py (hunk 2 above)
Comment (WillAyd) [reply pandas/tests/io/excel/test_writers.py]: Can you take a look at test_readers to see how we handle this? Can ideally ensure we parametrize with the appropriate writer / extension combinations up front; see `engine_and_read_ext` in test_readers

File: pandas/io/excel/_odswriter.py (hunk 3 above)
Comment (WillAyd) [reply pandas/io/excel/_odswriter.py]: Hmm still not clear on why this is needed and it slightly different from the other writers; can you try to remove?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_odswriter.py (hunk 3 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: Without this the following tests fail:
```
FAILED pandas/tests/io/excel/test_writers.py::TestRoundTrip::test_excel_multindex_roundtrip[1-1-True-False-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestRoundTrip::test_excel_multindex_roundtrip[3-1-True-False-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_roundtrip_indexlabels[True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_roundtrip_indexlabels[False-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_roundtrip_indexname[True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_roundtrip_indexname[False-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex[True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex[False-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex_nan_label[True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex_nan_label[False-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex_cols[False-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex_dates[True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_to_excel_multiindex_dates[False-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_010_hemstring[True-1-1-True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_010_hemstring[True-1-2-True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_010_hemstring[True-1-3-True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_010_hemstring[False-1-1-True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_010_hemstring[False-1-2-True-odf-.ods]
FAILED pandas/tests/io/excel/test_writers.py::TestExcelWriter::test_excel_010_hemstring[False-1-3-True-odf-.ods]
```

File: pandas/tests/io/excel/test_writers.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_writers.py]: I have now implemented an automatic file type recognizer for byte streams. Currently the only choice to be made is between 'xlrd' and 'odf', so it is quite easy.
For the other test I have improved the extension recognition to more types of `path`.
Plus I have fixed the parameterization of the tests that depend on the various external Excel read and write packages.

**Comment by alimcmaster1** [comment]
Can you merge master - thanks!

**Comment by roberthdevries** [comment]
Rebased to master

File: pandas/io/excel/_base.py [hunk 4]
```diff
@@ -809,17 +819,24 @@ class ExcelFile:
         "pyxlsb": _PyxlsbReader,
     }
 
-    def __init__(self, io, engine=None):
+    def __init__(self, path_or_io, engine=None):
         if engine is None:
             engine = "xlrd"
+            if isinstance(path_or_io, IOBase):
```
Comment (WillAyd) [inline pandas/io/excel/_base.py:825]: I don’t think we need to do this - are we doing this for xls vs .xlsx? Maybe better as a separate PR if you feel strongly and I’m wrong about the existing file types

File: pandas/io/excel/_base.py [hunk 5]
```diff
@@ -809,17 +819,24 @@ class ExcelFile:
...
+                if _is_ods_stream(path_or_io):
+                    engine = "odf"
+            else:
+                ext = os.path.splitext(str(path_or_io))[-1]
+                if ext == ".ods":
```
Comment (WillAyd) [inline pandas/io/excel/_base.py:830]: Rather than this should register the writer globally

https://github.com/pandas-dev/pandas/blob/120e9d9fae447be2e1cea3fd83a1f19501972cca/pandas/core/config_init.py#L576

File: pandas/io/excel/_odfreader.py (hunk 1 above)
Comment (WillAyd) [reply pandas/io/excel/_odfreader.py:3]: Can you just add a quick what’s new for this too?

**Review by WillAyd — CHANGES_REQUESTED** [review]
OK looking pretty good - thanks

File: pandas/io/excel/_base.py [hunk 6]
```diff
@@ -781,14 +778,27 @@ def close(self):
...
+
 class ExcelFile:
     """
     Class for parsing tabular excel sheets into DataFrame objects.
     Uses xlrd. See read_excel for more documentation
 
     Parameters
     ----------
-    io : str, path object (pathlib.Path or py._path.local.LocalPath),
+    path_or_io : str, path object (pathlib.Path or py._path.local.LocalPath),
```
Comment (jreback) [inline pandas/io/excel/_base.py:801]: path

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (jreback) [reply pandas/io/excel/_base.py:825]: I agree can you revert this

File: pandas/io/excel/_odfreader.py [hunk 7]
```diff
@@ -148,6 +150,9 @@ def _is_empty_row(self, row) -> bool:
     def _get_cell_value(self, cell, convert_float: bool) -> Scalar:
         from odf.namespaces import OFFICENS
 
+        if str(cell) == "#N/A":
```
Comment (jreback) [inline pandas/io/excel/_odfreader.py:153]: are there tests for NA handling?

File: pandas/io/excel/_odswriter.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+class _ODSWriter(ExcelWriter):
+    engine = "odf"
+    supported_extensions = (".ods",)
+
+    def __init__(self, path, engine=None, encoding=None, mode="w", **engine_kwargs):
+        from odf.opendocument import OpenDocumentSpreadsheet
+
+        engine_kwargs["engine"] = engine
+
+        if mode == "a":
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:20]: test for this?

File: pandas/io/excel/_odswriter.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    def save(self):
+        """
+        Save workbook to disk.
+        """
+        for sheet in self.sheets.values():
+            self.book.spreadsheet.addElement(sheet)
+        return self.book.save(self.path)
+
+    def write_cells(
+        self, cells, sheet_name=None, startrow=0, startcol=0, freeze_panes=None
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:37]: can you type signatures whereever possible & provide full doc-strings

File: pandas/io/excel/_odswriter.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            rows[cell.row].addElement(tc)
+            col_count[cell.row] += 1
+            p = P(text=pvalue)
+            tc.addElement(p)
+
+        # add all rows to the sheet
+        for row_nr in range(max(rows.keys()) + 1):
+            wks.addElement(rows[row_nr])
+
+    def _make_table_cell_attributes(self, cell):
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:83]: type & doc-string

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_base.py (hunk 6 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py:801]: This can be a path or a file-like object. The current doc-string is wrong.

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py:825]: No, I do this because of .xls* vs. .ods, and no it cannot be reverted unless tons of tests fail.

File: pandas/io/excel/_odswriter.py (hunk 8 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py:20]: not yet 😃

File: pandas/io/excel/_base.py (hunk 5 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py:830]: The weird thing is that this stuff does not seem to be used anywhere. Correct me if I'm wrong. I added a similar bit for the OpenOffice file format, but it did not seem to be called/tested anywhere.

**Comment by jreback** [comment]
@roberthdevries can you merge master and will look again.

File: doc/source/whatsnew/v1.1.0.rst [hunk 11]
```diff
@@ -288,6 +288,7 @@ Other enhancements
 - :meth:`HDFStore.put` now accepts `track_times` parameter. Parameter is passed to ``create_table`` method of ``PyTables`` (:issue:`32682`).
 - Make :class:`pandas.core.window.Rolling` and :class:`pandas.core.window.Expanding` iterable（:issue:`11704`)
 - Make ``option_context`` a :class:`contextlib.ContextDecorator`, which allows it to be used as a decorator over an entire function (:issue:`34253`).
+- :meth:`DataFrame.to_excel` can now also generate OpenOffice spreadsheet (.ods) files (:issue:`27222`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:291]: generate -> write

File: pandas/io/excel/_base.py (hunk 6 above)
Comment (jreback) [reply pandas/io/excel/_base.py:801]: kk, rename to path_or_buffer

File: pandas/io/excel/_base.py [hunk 12]
```diff
@@ -809,17 +819,24 @@ class ExcelFile:
         "pyxlsb": _PyxlsbReader,
     }
 
-    def __init__(self, io, engine=None):
+    def __init__(self, path_or_io, engine=None):
```
Comment (jreback) [inline pandas/io/excel/_base.py:822]: use path_or_buffer

File: pandas/io/excel/_odfreader.py (hunk 7 above)
Comment (jreback) [reply pandas/io/excel/_odfreader.py:153]: do these exist?

File: pandas/io/excel/_odswriter.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+from pandas.io.excel._base import ExcelWriter
+from pandas.io.excel._util import _validate_freeze_panes
+
+
+class _ODSWriter(ExcelWriter):
+    engine = "odf"
+    supported_extensions = (".ods",)
+
+    def __init__(self, path, engine=None, encoding=None, mode="w", **engine_kwargs):
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:15]: can you type as much as possible here

File: pandas/io/excel/_odswriter.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+
+        if mode == "a":
+            raise ValueError("Append mode is not supported with odf!")
+
+        super().__init__(path, mode=mode, **engine_kwargs)
+
+        self.book = OpenDocumentSpreadsheet()
+        self._style_dict: Dict[str, str] = {}
+
+    def save(self):
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:28]: can you type as much as possible (e.g. -> None)

**Review by jreback — CHANGES_REQUESTED** [review]
if you can type and doc-string as much as possible (could type as a follow up if @WillAyd ok with this)

**Comment by jreback** [comment]
can you rebase and update

File: pandas/io/excel/_odswriter.py (hunk 14 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: added type info

File: pandas/io/excel/_odswriter.py (hunk 13 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: added type info

File: pandas/io/excel/_base.py (hunk 12 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py]: changed as you suggested

File: doc/source/whatsnew/v1.1.0.rst (hunk 11 above)
Comment (roberthdevries) [reply doc/source/whatsnew/v1.1.0.rst]: Changed according to your suggestion

File: pandas/io/excel/_odswriter.py (hunk 10 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: added type info + docstrings

File: pandas/io/excel/_odfreader.py (hunk 7 above)
Comment (roberthdevries) [reply pandas/io/excel/_odfreader.py:153]: Yes: test_readers.py::TestReaders::test_excel_cell_error_na

File: pandas/io/excel/_base.py (hunk 6 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py]: Changed

File: pandas/io/excel/_base.py [hunk 15]
```diff
@@ -781,6 +778,19 @@ def close(self):
         return self.save()
 
 
+def _is_ods_stream(stream):
```
Comment (jreback) [inline pandas/io/excel/_base.py:781]: can you type this function inputs & outputs & add a doc-string

File: pandas/io/excel/_odswriter.py (hunk 8 above)
Comment (jreback) [reply pandas/io/excel/_odswriter.py:22]: did you update this?

File: pandas/io/excel/_odswriter.py [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+    def save(self) -> None:
+        """
+        Save workbook to disk.
+        """
+        for sheet in self.sheets.values():
+            self.book.spreadsheet.addElement(sheet)
+        self.book.save(self.path)
+
+    def write_cells(
+        self,
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:39]: can you type these

File: pandas/io/excel/_odswriter.py [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+            rows[cell.row].addElement(tc)
+            col_count[cell.row] += 1
+            p = P(text=pvalue)
+            tc.addElement(p)
+
+        # add all rows to the sheet
+        for row_nr in range(max(rows.keys()) + 1):
+            wks.addElement(rows[row_nr])
+
+    def _make_table_cell_attributes(self, cell) -> Dict[str, object]:
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:90]: can you type here

File: pandas/io/excel/_odswriter.py [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+        attributes: Dict[str, object] = {}
+        style_name = self._process_style(cell.style)
+        if style_name is not None:
+            attributes["stylename"] = style_name
+        if cell.mergestart is not None and cell.mergeend is not None:
+            attributes["numberrowsspanned"] = max(1, cell.mergestart)
+            attributes["numbercolumnsspanned"] = cell.mergeend
+        return attributes
+
+    def _make_table_cell(self, cell) -> Tuple[str, object]:
```
Comment (jreback) [inline pandas/io/excel/_odswriter.py:112]: and here, etc

File: pandas/tests/io/excel/test_writers.py [hunk 19]
```diff
@@ -48,11 +48,19 @@ def set_engine(engine, ext):
...
+        pytest.param(".xlsm", marks=[td.skip_if_no("openpyxl"), td.skip_if_no("xlrd")]),
+        pytest.param(".xls", marks=[td.skip_if_no("xlwt"), td.skip_if_no("xlrd")]),
+        pytest.param(
+            ".xlsx", marks=[td.skip_if_no("xlsxwriter"), td.skip_if_no("xlrd")]
+        ),
+        pytest.param(".ods", marks=td.skip_if_no("odf")),
+    ],
+)
 class TestRoundTrip:
-    @td.skip_if_no("xlwt")
```
Comment (jreback) [inline pandas/tests/io/excel/test_writers.py:54]: so these now work even when these are not installed? or skip appropriately?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/excel/test_writers.py (hunk 19 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_writers.py:54]: The skip_if_no is moved to the parametrize decorator, where is is present for the applicable file types.

File: pandas/io/excel/_odswriter.py (hunk 17 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py]: Done, hopefully correct

File: pandas/io/excel/_odswriter.py (hunk 18 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py:114]: This one is a bit more difficult as the type is an odf.table.TableCell. As the odf package is is not installed on most test environments, this type cannot be added or the test will fail.

File: pandas/io/excel/_base.py (hunk 15 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py]: added

File: pandas/io/excel/_odswriter.py (hunk 16 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py:40]: done

File: pandas/io/excel/_odswriter.py (hunk 8 above)
Comment (roberthdevries) [reply pandas/io/excel/_odswriter.py:23]: A test has been added: pandas/tests/io/excel/test_odswriter.py

File: pandas/io/excel/_odfreader.py (hunk 1 above)
Comment (roberthdevries) [reply pandas/io/excel/_odfreader.py:3]: Added

**Review by jreback — APPROVED** [review]
lgtm. can always open an issue for followups if needed. cc @WillAyd

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Great thanks @roberthdevries

**Comment by roberthdevries** [comment]
You're welcome
